### ISSUE TITLE
Composite Intersection of optional field doesn't preserve optionality

### ISSUE THREAD

**Issue by Abe27342** [sinclairzx81/typebox#482]
The logic which intersects each key of a composite type appears to not preserve optionality--codebox [here](https://codesandbox.io/s/young-sunset-tp62js) with the crux of the issue demonstrated by:

```typescript
import { Type } from "@sinclair/typebox";

const foo = Type.Object({
  a: Type.Optional(Type.Number())
});

const baz = Type.Composite([foo, foo]);

console.log("baz schema", JSON.stringify(baz)); // shows "a" as a required property
```

Seems like the result of creating the `baz` composite type should be able to preserve optionality of a given `key` if all of the property schema in the intersection are optional.

This isn't a huge issue for me: I only noticed it because some schema I use is built up from a lot of mixin-like patterns, and one of them was included twice (which is bad anyway, and how I intend to fix my use case).
Figured I would report in case it's giving anyone else trouble, or if it's an interesting case to think about for the removal of `Composite` in favor of `Mapped`.

**Comment by sinclairzx81** [comment]
@Abe27342 Hi! Thanks for reporting! :)

Yeah, this has been a difficult one to solve, but a fix is planned for when I next get some time to work through it. The problem actually relates to Indexed Access Types and the 0.28.0 handling of overlapping properties with varying modifiers. It's a bit of a complex problem, but it relates specifically to TS rules when intersecting overlapping property keys with varying modifiers; as well as some of the challenges expressing those rules through TB inference mapping.

Will draft the problem below (for future reference), but may not be able to resolve in the short term. As a workaround, you can avoid the problem by ensuring no property keys overlap on calls to Composite, or if property keys do overlap, they have no modifiers applied to their type. Alternatively, using Intersect will work on all cases.

### Problem

At a high level, the following demonstrates the issue (note the type `J` below) as Composite uses Indexed Access Types during property enumeration.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKZwL5wGZQQhwDkAAgM7AB2AxgDYCGwUA9DOhgEYQAeJAWABQw2hGoV4AQTgBeFJwB0AeS4ArDLRgAKJIwBcCzCrAxg4xvW1pjAOQCuILhijaAlG5yfWrOBABuLkxgYDQA5nBghJiwwBgUouKScABCckYYKuqaOnqGNpnKpubUltZKDk4u7p7YbsKJEvAAwukFis1EkFQwGNoA2lIANKkAup5wcD4oqhpaADwIwpMrq2vrG5vTkwYoAJLUvVAUOfP9y5uXVxvbK8hFZhb088iVzlAAfEMX179Xt5N7sUni83i4Pj8-lC1rdRh8ANyQ6HI6bYCEiITTRTYuAAd2AMAAFnAiVgohAYhw4AADRjUuAAExcwECDPwhGINCZvAwbMYtFo8Qo0DgDSEYiacD2bSUByOJy0A2GYwmooxEuSACkZcYDtztHsRv0SIwSOMYb5+TB7JZ8nKXAqYGcgY9Ss9Xo53l8ZsC3aDPeC4UjkVCfMGQ79phheJgtPkHiVLC97cdTv0PVUoCMM+84R8gA)

```typescript
import { Type } from '@sinclair/typebox'

const A = Type.Object({ a: Type.Optional(Type.Number()) }) // overlapping properties
const B = Type.Object({ a: Type.Optional(Type.Number()) })

const C = Type.Composite([A, B])   // TObject<{
                                   //   a: TIntersect<[
                                   //     TOptional<TNumber>,
                                   //     TOptional<TNumber>
                                   //   ]>;
                                   // }>

// ... with the property `a` derived from indexed accessor 

const I = Type.Intersect([A, B])   

const J = Type.Index(I, ['a'])     // actual: TIntersect<[TOptional<TNumber>, TOptional<TNumber>]>
                                   //
                                   // expect: TOptional<TIntersect<[TNumber, TNumber]>>
```
### Conditions

The conditions for remapping are that the optional should only be remapped when ALL sub types of overlapping property are optional (in all other cases, the optional is omitted). The following is the TS behavior for this, however the TB type inference mapping to replicate this is quite expensive (and there are some concerns about pushing out TS recursive depth heuristics performing the mapping)

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/C4TwDgpgBAogbgQwDYFcHAgHgCoD4oC8U2UEAHhgHYAmAzlAJaUBmEATlAKpQD8UA3lADaAaUaUoAawggA9sy4BdAFxdRigNxQAvlFWUIcdgFgAUGcEB6S1ADGsmg2AMHUAIxmoUUJCgBBQgEoBH0UAFsAI3YdT29waAAhQMEQqEpwqI5tWJ9oAGFA+GQ0DEwAgDIoBPwoazjfAqIU0Mj2LWzTDqsbe0dnVwAmHPj-ZOCeFsyY0y9cqrHU9Nas4YbCxFR0LAqqmrq5xqDFjLbprtqeh2onFwkAZigAWig2FCQIVQckEGCwMG+oLIwP1KMhGAo-AAZSGAoxsJAIP5MADmUFoKAi9Qg9AQbGgQJByFW0ACTXGk2iHVmIySZIQEzSJxWMyxUEORU2pR21S8+xGhxSDKWUwAPlAUDQIMwmBBqO0zNogA)

```typescript
type Evaluate<T> = T extends infer U ? { [K in keyof U]: U[K]; } : never

{ // condition 1
  type A = { a: number }
  type B = { a: number }
  type C = Evaluate<A & B>  // type C = { a: number; }
}
{ // condition 2
  type A = { a?: number }
  type B = { a: number }
  type C = Evaluate<A & B>  // type C = { a: number; }
}
{ // condition 3 - rule: only apply optional if ALL overlapping sub types are optional
  type A = { a?: number }
  type B = { a?: number }
  type C = Evaluate<A & B>  // type C = { a?: number | undefined; }
}
```
These considerations would be valid for `readonly` mapping also (which requires looking into). 

### Current

So, on the 0.28.0 release, I've opted to keep the runtime mapping and inference mapping in sync, meaning if you hover the return type of Composite, the type inference matches the runtime mapping. This was done to help reveal the underlying mapping and what needs to be done to resolve (and to help me get my bearings on issue when I next look into it :D)

Will aim to have a resolution to this issue when the 0.29.0 release rolls around (where I'll be taking another look at Mapped types)
All the best!
S

**Comment by Abe27342** [comment]
Thanks for the details and roadmap! Sounds great to me. Your suggested workaround makes sense in the meantime.

**Comment by sinclairzx81** [comment]
@Abe27342 Heya,

Have just published a fix for this on `0.29.2`. Information about the update can be found [here](https://github.com/sinclairzx81/typebox/blob/master/changelog/0.29.2.md). Note that this update only applies to `optional`. I may look into preserving `readonly` modifiers in subsequent revisions.

[TypeScript Link Here](https://www.typescriptlang.org/play?ts=5.1.3#code/JYWwDg9gTgLgBAbzgFQJ5gKZwL5wGZQQhwDkAAgM7AB2AxgDYCGwUA9DOhgEYQAeJAWABQw2hGoV4AMQgQ4AXhScAdAHkuAKwy0YACgTC4cRgC4lmNWBjBxjerrQWAcgFcQXDFF0BKb8Ox+IkJiEvAAQoxQCuYYygDCRJBUMBi6ANoyEAA0cJkAut5wrKxwIZJwEVGKBkJGdfUNjU3N9cV1HJhmAEQQmtowXVmGLSOjzW1GYISYsMAYFGY1Y8srE3WmiMMr2+MljXb0qnhmaUs75w1rTR0Y3dRuHlBdWxcXV3XYOWevb3vNN3cHp5nrUfr8mtg8i8was-kZsNCYWM1gjQUjYThhKJxOVkNFHLEAJLUAAmGF4ukqOTSJEYJAKTTaNxQ+NUVhs1DsAB5kMSUlAKP0uWlkK53J4cqKgVA8gA+WVAA)

```typescript
const Foo = Type.Object({
  a: Type.Optional(Type.Number())
})

const Bar = Type.Composite([Foo, Foo]) // const Bar = {
                                       //   type: "object",
                                       //   properties: {
                                       //     a: {
                                       //       allOf: [{
                                       //          type: "number"
                                       //       }, {
                                       //          type: "number"
                                       //       }]
                                       //     }
                                       //   }
                                       // }

const T = Type.Index(Bar, ['a'])       // type T = TOptional<TIntersect<[TNumber, TNumber]>>
```
Have tested inference from TS 4.1.5 and up and should be ok.

**Comment by sinclairzx81** [comment]
@Abe27342 Hey might close up this issue as this inference problem should be resolved on `0.29.2` (with a small keyof fix on `0.29.3`) 

Managed to get this though on the 0.29.x release after all (this revision was fast tracked due to a instantiation issue on TS 5.1.6)). Fairly confident with the implementation, but will keep an eye on any problems coming through over the next few weeks. Should be good though.

All the best!
S
